// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: moego/business/staff/v1/staff_service.proto

package staffpb

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on GetStaffRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetStaffRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetStaffRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetStaffRequestMultiError, or nil if none found.
func (m *GetStaffRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetStaffRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetStaffRequestMultiError(errors)
	}

	return nil
}

// GetStaffRequestMultiError is an error wrapping multiple validation errors
// returned by GetStaffRequest.ValidateAll() if the designated constraints
// aren't met.
type GetStaffRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetStaffRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetStaffRequestMultiError) AllErrors() []error { return m }

// GetStaffRequestValidationError is the validation error returned by
// GetStaffRequest.Validate if the designated constraints aren't met.
type GetStaffRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetStaffRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetStaffRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetStaffRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetStaffRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetStaffRequestValidationError) ErrorName() string { return "GetStaffRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetStaffRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetStaffRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetStaffRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetStaffRequestValidationError{}

// Validate checks the field values on ListStaffsRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListStaffsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListStaffsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListStaffsRequestMultiError, or nil if none found.
func (m *ListStaffsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListStaffsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPagination()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListStaffsRequestValidationError{
					field:  "Pagination",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListStaffsRequestValidationError{
					field:  "Pagination",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPagination()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListStaffsRequestValidationError{
				field:  "Pagination",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ListStaffsRequestMultiError(errors)
	}

	return nil
}

// ListStaffsRequestMultiError is an error wrapping multiple validation errors
// returned by ListStaffsRequest.ValidateAll() if the designated constraints
// aren't met.
type ListStaffsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListStaffsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListStaffsRequestMultiError) AllErrors() []error { return m }

// ListStaffsRequestValidationError is the validation error returned by
// ListStaffsRequest.Validate if the designated constraints aren't met.
type ListStaffsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListStaffsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListStaffsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListStaffsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListStaffsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListStaffsRequestValidationError) ErrorName() string {
	return "ListStaffsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListStaffsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListStaffsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListStaffsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListStaffsRequestValidationError{}

// Validate checks the field values on ListStaffsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListStaffsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListStaffsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListStaffsResponseMultiError, or nil if none found.
func (m *ListStaffsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListStaffsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for NextPageToken

	for idx, item := range m.GetStaffs() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListStaffsResponseValidationError{
						field:  fmt.Sprintf("Staffs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListStaffsResponseValidationError{
						field:  fmt.Sprintf("Staffs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListStaffsResponseValidationError{
					field:  fmt.Sprintf("Staffs[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListStaffsResponseMultiError(errors)
	}

	return nil
}

// ListStaffsResponseMultiError is an error wrapping multiple validation errors
// returned by ListStaffsResponse.ValidateAll() if the designated constraints
// aren't met.
type ListStaffsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListStaffsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListStaffsResponseMultiError) AllErrors() []error { return m }

// ListStaffsResponseValidationError is the validation error returned by
// ListStaffsResponse.Validate if the designated constraints aren't met.
type ListStaffsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListStaffsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListStaffsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListStaffsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListStaffsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListStaffsResponseValidationError) ErrorName() string {
	return "ListStaffsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListStaffsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListStaffsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListStaffsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListStaffsResponseValidationError{}
